### PR TITLE
raftstore/apply: tiny refactor

### PR DESCRIPTION
**Pull request by BusyJay**
Make `handle_raft_committed_entries` handle meta processing, so it can be reused in merge later.

This pr also breaks apply into `prepare_for`, `commit`, `finish_for` and `flush` four stages to make things clear.

### CODE REVIEW

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -269,20 +346,21 @@ pub fn notify_stale_req(term: u64, cb: Callback) {
     cb.invoke_with_response(resp);
 }
 
-fn should_flush_to_engine(cmd: &RaftCmdRequest, wb_keys: usize) -> bool {
-    // When encounter ComputeHash cmd, we must flush the write batch to engine immediately.
+/// Check if a commit is needed to be issued before handle the command.
+fn should_commit_to_engine(cmd: &RaftCmdRequest, wb_keys: usize) -> bool {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: Can we use another name? There are too many `commit`.

Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: Do you have any suggestion?

Comment (zhangjinpeng1987) [reply src/raftstore/store/worker/apply.rs]: should_put/write_to_engine?

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -195,48 +194,126 @@ impl ApplyCallback {
...
         ApplyContext {
             host: host,
-            wb: WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE),
+            wb: None,
             cbs: MustConsumeVec::new("callback of apply context"),
+            apply_res: vec![],
             wb_last_bytes: 0,
             wb_last_keys: 0,
+            committed_count: 0,
+            sync_log_hint: false,
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: It is easy to confuse with `sync_log_hint` and `sync_log`.

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -195,48 +194,131 @@ impl ApplyCallback {
 
 struct ApplyContext<'a> {
     host: &'a CoprocessorHost,
-    wb: WriteBatch,
+    wb: Option<WriteBatch>,
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: What's the benefit of using an `Option` here?

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -195,48 +194,131 @@ impl ApplyCallback {
...
     cbs: MustConsumeVec<ApplyCallback>,
+    apply_res: Vec<ApplyRes>,
     wb_last_bytes: u64,
     wb_last_keys: u64,
+    committed_count: usize,
+    // `sync_log_hint` is a global hint indicate that whether synchronize wal
+    // when data is written to kv engine.
+    sync_log_hint: bool,
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: Seems the only use case is that, if `sync_log && sync_log_hint`, we need to synchronize WAL, can we combine them together?

File: src/raftstore/store/worker/apply.rs [hunk 5]
```diff
@@ -269,20 +351,21 @@ pub fn notify_stale_req(term: u64, cb: Callback) {
     cb.invoke_with_response(resp);
 }
 
-fn should_flush_to_engine(cmd: &RaftCmdRequest, wb_keys: usize) -> bool {
-    // When encounter ComputeHash cmd, we must flush the write batch to engine immediately.
+/// Check if a commit is needed to be issued before handle the command.
+fn should_write_to_engine(cmd: &RaftCmdRequest, wb_keys: usize) -> bool {
+    // When encounter ComputeHash cmd, we must commit the write batch to engine immediately.
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: s/commit/write

File: src/raftstore/store/worker/apply.rs [hunk 6]
```diff
@@ -269,20 +351,21 @@ pub fn notify_stale_req(term: u64, cb: Callback) {
...
     if cmd.has_admin_request()
         && cmd.get_admin_request().get_cmd_type() == AdminCmdType::ComputeHash
     {
         return true;
     }
 
-    // When write batch contains more than `recommended` keys, flush the batch to engine.
+    // When write batch contains more than `recommended` keys, commit the batch to engine.
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: s/commit/write

File: src/raftstore/store/worker/apply.rs (hunk 3 above)
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: To be able take the wb without drop context or creating a new `WriteBatch`.

File: src/raftstore/store/worker/apply.rs (hunk 4 above)
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: It's not easy when using bool. `sync_log_hint` is set by configuration, and `sync_log` is set depend on the content of entries being applied.

File: src/raftstore/store/worker/apply.rs [hunk 7]
```diff
@@ -195,48 +194,131 @@ impl ApplyCallback {
...
+    ///
+    /// This call is valid only when it's between a `prepare_for` and `finish_for`.
+    fn commit(&mut self, delegate: &mut ApplyDelegate, persistent: bool) {
+        if persistent {
+            delegate.write_apply_state(self.wb_mut());
+        }
+        delegate.update_metrics(self);
+        if persistent {
+            self.write_to_db(&delegate.engine);
+            if self.wb.is_none() {
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: Seems the same as `prepare_for`?

File: src/raftstore/store/worker/apply.rs [hunk 8]
```diff
@@ -195,48 +194,131 @@ impl ApplyCallback {
...
+    fn use_delete_range(mut self, use_delete_range: bool) -> ApplyContext<'a> {
+        self.use_delete_range = use_delete_range;
+        self
+    }
+
+    /// Prepare for applying entries for `delegate`.
+    ///
+    /// A general apply progress for a delegate is:
+    /// `prepare_for` -> `commit` -> `commit` -> `finish_for`.
+    /// After all delegates are handled, `flush` method should be called.
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: s/flush/write_to_db

File: src/raftstore/store/worker/apply.rs (hunk 2 above)
Comment (huachaohuang) [reply src/raftstore/store/worker/apply.rs]: Maybe we should rename `sync_log_hint` as `enable_sync_log` and `sync_log` as `should_sync_log`.

File: src/raftstore/store/worker/apply.rs [hunk 9]
```diff
@@ -195,48 +194,126 @@ impl ApplyCallback {
...
+
+    fn use_delete_range(mut self, use_delete_range: bool) -> ApplyContext<'a> {
+        self.use_delete_range = use_delete_range;
+        self
+    }
+
+    /// Prepare for applying entries for `delegate`.
+    ///
+    /// A general apply progress for a delegate is:
+    /// `prepare_for` -> `commit` -> `commit` -> `finish_for`.
```
Comment (hicqu) [inline src/raftstore/store/worker/apply.rs]: 2 `commit`?

File: src/raftstore/store/worker/apply.rs [hunk 10]
```diff
@@ -195,48 +194,126 @@ impl ApplyCallback {
...
 
-    pub fn mark_last_bytes_and_keys(&mut self) {
-        self.wb_last_bytes = self.wb.data_size() as u64;
-        self.wb_last_keys = self.wb.count() as u64;
+    /// Commit all changes have done for delegate. `persistent` indicates whether
+    /// write the changes into rocksdb.
+    ///
+    /// This call is valid only when it's between a `prepare_for` and `finish_for`.
+    fn commit(&mut self, delegate: &mut ApplyDelegate, persistent: bool) {
+        if persistent {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/apply.rs]: I prefer to merge this if and the below if, although the apply state will not be included in statistics.

Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: That's the exact reason why they should not be merged.

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by BusyJay** [comment]
PTAL

**Comment by BusyJay** [comment]
@overvenus PTAL, /run-integration-tests

**Pull request merged**